azurerm/internal/services/managedapplication/resource_arm_managed_application_definition.go

```diff
@@ -0,0 +1,313 @@
+				DiffSuppressFunc: structure.SuppressJsonDiff,
+				ConflictsWith:    []string{"package_file_uri"},
+			},
+
+			"package_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"package_file_uri": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.IsURLWithHTTPorHTTPS,
+			},
+
+			"tags": tags.Schema(),
+		},
+	}
+}
+
+func resourceArmManagedApplicationDefinitionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).ManagedApplication.ApplicationDefinitionClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroupName := d.Get("resource_group_name").(string)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroupName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing Managed Application Definition Name %q (Resource Group %q): %+v", name, resourceGroupName, err)
+			}
+		}
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_managed_application_definition", *existing.ID)
+		}
+	}
+
+	location := azure.NormalizeLocation(d.Get("location").(string))
+	authorizations := d.Get("authorization").(*schema.Set).List()
+	displayName := d.Get("display_name").(string)
+	description := d.Get("description").(string)
+	packageEnabled := d.Get("package_enabled").(bool)
+	lockLevel := d.Get("lock_level").(string)
+	t := d.Get("tags").(map[string]interface{})
+
+	parameters := managedapplications.ApplicationDefinition{
+		Location: utils.String(location),
+		ApplicationDefinitionProperties: &managedapplications.ApplicationDefinitionProperties{
+			Authorizations: expandArmManagedApplicationDefinitionAuthorization(authorizations),
+			Description:    utils.String(description),
+			DisplayName:    utils.String(displayName),
+			IsEnabled:      utils.Bool(packageEnabled),
+			LockLevel:      managedapplications.ApplicationLockLevel(lockLevel),
+		},
+		Tags: tags.Expand(t),
+	}
+
+	if v, ok := d.GetOk("create_ui_definition"); ok {
+		parameters.CreateUIDefinition = utils.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("main_template"); ok {
+		parameters.MainTemplate = utils.String(v.(string))
+	}
+
+	if (parameters.CreateUIDefinition != nil && parameters.MainTemplate == nil) || (parameters.CreateUIDefinition == nil && parameters.MainTemplate != nil) {
+		return fmt.Errorf("`create_ui_definition` and `main_template` should be set or not set together")
+	}
+
+	if v, ok := d.GetOk("package_file_uri"); ok {
+		parameters.PackageFileURI = utils.String(v.(string))
+	}
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroupName, name, parameters)
+	if err != nil {
+		return fmt.Errorf("creating Managed Application Definition %q (Resource Group %q): %+v", name, resourceGroupName, err)
```

Comment:
Can we express the meaning of error message explicitly? (same for other error messages)

```suggestion
		return fmt.Errorf("failed to create Managed Application Definition %q (Resource Group %q): %+v", name, resourceGroupName, err)
```

Reply:
updated